### DISCUSSION TITLE
"Connection refused" when executing Function

### DISCUSSION THREAD

**Discussion by zdanl** [appwrite/appwrite#4854] [Q&A] [answered]
### 👟 Reproduction steps

Fresh, brand-new, as per appwrite.io instructions, fetched & installed Docker instance, simple Python Function, Fails.

![appwrite_function](https://user-images.githubusercontent.com/114028070/207187030-2e8c75f3-5e0e-4c8c-ab9f-faebd19e0f3e.png)

### 👍 Expected behavior

Execution of the Python Function.

### 👎 Actual Behavior

Python Function fails.

![appwrite_function](https://user-images.githubusercontent.com/114028070/207186944-e2387e87-6d0f-452f-8bf2-203972c35466.png)

### 🎲 Appwrite version

Version 1.1.x

### 💻 Operating system

Linux

### 🧱 Your Environment

I use Scaleway.com a hoster with the newest Ubuntu 22 and Serverside Python3.9 Python Functions and a Clientside ReactJS/NextJS Setup.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by zdanl** [comment]
The product promises to take care of requirements.txt .... So it should. When you reduce the requirements to zero, and run a simple time.time(), that works. 

The output of installing the web3 requirement is incomplete, so is the output of what is actually going wrong. Both on shell, where a customer shouldn't have to go (the whole deal about AppWrite is so that you don't have to care about the Backend other than implementing into your Frontend, which in times of TypeScript/ReactJS/NextJS is hard enough....) and in the web interface. 

This should be worked on.

I tried to bash into the executor container, but I don't know how it works yet, there is no bash. 

Any advise on how to debug that / any guidance on how the system works so I can fix some things for you?
It's probably still faster than building my own Backend in NodeJS / MongoDB.

**Comment by stnguyen90** [comment, marked as answer]
Connection refused typically occurs when something is wrong with the function code preventing it from running properly.

I highly recommend following the [Functions docs](https://appwrite.io/docs/functions) for initializing and deploying functions. For me, I added `web3` to my `requirements.txt` and it showed up in the build logs:

![image](https://user-images.githubusercontent.com/1477010/207403960-509114de-102b-4706-bc1a-fc9ebe5a8756.png)

When executing the following code:

```python
from web3 import EthereumTesterProvider, Web3

def main(req, res):
  try:
    w3 = Web3(EthereumTesterProvider())
    return res.json({
      "success": True,
      "isConnected": w3.isConnected(),
    })
  except Exception as e:
    return res.json({
      "success": False,
      "message": str(e)
    })

```

 I still got an error. However, after moving the import like:
 
 ```python
def main(req, res):
  try:
    from web3 import EthereumTesterProvider, Web3
    w3 = Web3(EthereumTesterProvider())
    return res.json({
      "success": True,
      "isConnected": w3.isConnected(),
    })
  except Exception as e:
    return res.json({
      "success": False,
      "message": str(e)
    })
```

I got an error indicating:

```json
{"message":"No module named 'eth_tester'","success":false}
```

After including `eth-tester` in my `requirements.txt` file the test worked:

```json
{"isConnected":true,"success":true}
```

**Comment by zdanl** [comment]
With respect to your work here, there is a more pressing issue. 

I followed your instructions and deployed on shell, but database createDoc only works for 1 Document. the unique() ID.unique() isn't working. The server is also not accepting random strings. 

EDIT: Please don't force me to go into detail, you know what I mean. upsert() and such

**Reply by stnguyen90** [reply]
> I followed your instructions and deployed on shell, but database createDoc only works for 1 Document. the unique() ID.unique() isn't working. The server is also not accepting random strings.

The following code:

```python
from appwrite.client import Client
from appwrite.id import ID
from appwrite.services.databases import Databases
from web3 import EthereumTesterProvider, Web3


def main(req, res):
  client = Client()
  databases = Databases(client)

  try:
    w3 = Web3(EthereumTesterProvider())
    is_connected = w3.isConnected()

    if not req.variables.get('APPWRITE_FUNCTION_ENDPOINT') or not req.variables.get('APPWRITE_FUNCTION_API_KEY'):
      print('Environment variables are not set. Function cannot use Appwrite SDK.')
    else:
      (
      client
        .set_endpoint(req.variables.get('APPWRITE_FUNCTION_ENDPOINT', None))
        .set_project(req.variables.get('APPWRITE_FUNCTION_PROJECT_ID', None))
        .set_key(req.variables.get('APPWRITE_FUNCTION_API_KEY', None))
        .set_self_signed(True)
      )

    database_id = 'default'
    collection_id = 're'
    document_id = ID.unique()
    data = {
      "isConnected": is_connected
    }
    doc = databases.create_document(
      database_id,
      collection_id,
      document_id,
      data
    )

    return res.json({
      "success": True,
      "isConnected": is_connected,
      "id": doc['$id']
    })
  except Exception as e:
    return res.json({
      "success": False,
      "message": str(e)
    })

```

works for me:

![image](https://user-images.githubusercontent.com/1477010/207928221-affbb217-cd89-4c18-b244-1a79b747a022.png)

What kind of error are you seeing? Don't forget to create the `APPWRITE_FUNCTION_ENDPOINT` and `APPWRITE_FUNCTION_API_KEY` variables. Another problem some people run into is with their firewall. You could test with the firewall disabled to verify. If it is a firewall problem, the solution is to add a firewall rule for your Appwrite ports.

**Comment by zdanl** [comment]
Perhaps we should just get used to returning errors in HTTP JSON. 

`appwrite execute function` is not going through tho. So jumping to the web interface (Console[tm]) and removing the Anonymous (Guest) user Session by hand all the time, I know this can be automated. ...... , is not ... good.

**Reply by stnguyen90** [reply]
Sorry, can you clarify where the Anonymous (Guest) user Session comes into play?

**Comment by zdanl** [comment]
Bot Detection, more like Fingerprinting, for a high-assurance NFT platform with centralized socket into decentralized risk.

Basically, a Site Visitor that is not authenticated and has no wallet connection either, that needs to be server-side registered.

**Reply by stnguyen90** [reply]
Is this supposed to be a reply on one of the existing threads? Apologies, I also don't quote understand the comment and context.

**Comment by zdanl** [comment]
Oh ethereum wallet worked for 2 days. I'm on other ethical functions that I won't share for morality audit

________________________________
